WebDriver: unable to run tests using debug safaridriver, hitting ASSERT(!decoder.destinationID())

burg opened the pull request:
#### 0097369444dd4fe58699fb18c209385ef4c9f997
<pre>
WebDriver: unable to run tests using debug safaridriver, hitting ASSERT(!decoder.destinationID())
<a href="https://bugs.webkit.org/show_bug.cgi?id=282055">https://bugs.webkit.org/show_bug.cgi?id=282055</a>

Reviewed by Alex Christensen and Abrar Rahman Protyasha.

We hit this assertion when dispatching WebAutomationSessionProxy::Messages::EvaluateJavaScriptFunction.
This code was refactored on the UIProcess-side to call sendToProcessContainingFrame(), which includes
a destinationID. However, a destinationID is not expected because WebAutomationSessionProxy is a singleton
in WebProcess which outlives specific pages or frames.

To work around this, introduce a new sendToProcessContainingFrame() funnel method which does not include
the destinationID when dispatching the two destinationID-less messages to WebAutomationSessionProxy.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::evaluateJavaScriptFunction):
(WebKit::WebAutomationSession::takeScreenshot):

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::sendToProcessContainingFrameWithoutDestinationIdentifier): Added.
Adjust template instantiation macros so that the affected WebAutomationSessionProxy messages are sent
without a destinationId.

Canonical link: <a href="https://commits.webkit.org/286231@main">https://commits.webkit.org/286231@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/96c10c4d275c5f103af6b65287e21728fe8afde1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/73470 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52899 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/26278 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/77734 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/24708 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62030 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/684 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/57732 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16171 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/76537 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47812 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63206 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38142 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/44425 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23039 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66228 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21042 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/79354 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/786 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/281 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66127 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/927 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63219 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/65404 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16571 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9263 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/7434 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/749 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/3492 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/779 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/765 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/785 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15194,6 +15194,16 @@ std::optional<IPC::AsyncReplyID> WebPageProxy::sendWithAsyncReplyToProcessContai
+void WebPageProxy::sendToProcessContainingFrameWithoutDestinationIdentifier(std::optional<FrameIdentifier> frameID, M&& message, OptionSet<IPC::SendOption> options)
+{
+    sendToWebPage(frameID,
+        [&message, options] (auto& targetPage) {
```
Style guide nit; no space between capture list and arg list: https://webkit.org/code-style-guidelines/#spacing-lambda-paren

burg committed:
WebDriver: unable to run tests using debug safaridriver, hitting ASSERT(!decoder.destinationID())
https://bugs.webkit.org/show_bug.cgi?id=282055

Reviewed by Alex Christensen and Abrar Rahman Protyasha.

We hit this assertion when dispatching WebAutomationSessionProxy::Messages::EvaluateJavaScriptFunction.
This code was refactored on the UIProcess-side to call sendToProcessContainingFrame(), which includes
a destinationID. However, a destinationID is not expected because WebAutomationSessionProxy is a singleton
in WebProcess which outlives specific pages or frames.

To work around this, introduce a new sendToProcessContainingFrame() funnel method which does not include
the destinationID when dispatching the two destinationID-less messages to WebAutomationSessionProxy.

* Source/WebKit/UIProcess/Automation/WebAutomationSession.cpp:
(WebKit::WebAutomationSession::evaluateJavaScriptFunction):
(WebKit::WebAutomationSession::takeScreenshot):

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::sendToProcessContainingFrameWithoutDestinationIdentifier): Added.
Adjust template instantiation macros so that the affected WebAutomationSessionProxy messages are sent
without a destinationId.

Canonical link: https://commits.webkit.org/286231@main
